File: 151051/Makovetskiy/task330+340+350/dc-rest/src/main/java/com/rmakovetskij/dc_rest/service/LabelService.java
package com.rmakovetskij.dc_rest.service;

import com.rmakovetskij.dc_rest.mapper.LabelMapper;
import com.rmakovetskij.dc_rest.model.Label;
import com.rmakovetskij.dc_rest.model.dto.requests.LabelRequestTo;
import com.rmakovetskij.dc_rest.model.dto.responses.LabelResponseTo;
import com.rmakovetskij.dc_rest.repository.interfaces.LabelRepository;
import lombok.AllArgsConstructor;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.cache.annotation.EnableCaching;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;
import java.util.Optional;

@Service
@AllArgsConstructor
@EnableCaching
public class LabelService {

    private final LabelRepository labelRepository;
    private final LabelMapper labelMapper;

    @CacheEvict(value = "labels", allEntries = true)
    public LabelResponseTo createLabel(LabelRequestTo labelRequestDto) {
        Label label = labelMapper.toEntity(labelRequestDto);
        label = labelRepository.save(label);
        return labelMapper.toResponse(label);
    }

    @Cacheable(value = "labels", key = "#id")
    public LabelResponseTo getLabelById(Long id) {
        Optional<Label> labelOptional = labelRepository.findById(id);
        return labelOptional.map(labelMapper::toResponse)
                .orElseThrow(() -> new RuntimeException("Label not found"));
    }

    @Cacheable(value = "labelsList")
    public List<LabelResponseTo> getAllLabels() {
        return labelRepository.findAll().stream()
                .map(labelMapper::toResponse)
                .toList();
    }
    @CacheEvict(value = "labels", key = "#id")
    public LabelResponseTo updateLabel(Long id, LabelRequestTo labelRequestDto) {
        Label existingLabel = labelRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Label not found"));

        existingLabel.setName(labelRequestDto.getName());

        labelRepository.save(existingLabel);
        return labelMapper.toResponse(existingLabel);
    }

    @CacheEvict(value = "labels", key = "#id")
    public void deleteLabel(Long id) {
        if (!labelRepository.existsById(id)) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Label not found");
        }
        labelRepository.deleteById(id);
    }

    public boolean existsById(Long id) {
        return labelRepository.existsById(id);
    }

}

File: 251003/Pankratiev/rest/src/main/kotlin/bsuir/dc/rest/service/LabelService.kt
package bsuir.dc.rest.service

import bsuir.dc.rest.dto.from.LabelRequestTo
import bsuir.dc.rest.dto.to.LabelResponseTo
import bsuir.dc.rest.mapper.toEntity
import bsuir.dc.rest.mapper.toResponse
import bsuir.dc.rest.repository.IssueRepository
import bsuir.dc.rest.repository.LabelRepository
import org.springframework.stereotype.Service

@Service
class LabelService(
    private val labelRepository: LabelRepository,
    private val issueRepository: IssueRepository,
) {
    fun createLabel(labelRequestTo: LabelRequestTo): LabelResponseTo {
        val label = labelRequestTo.toEntity()
        val savedLabel = labelRepository.save(label)
        return savedLabel.toResponse()
    }

    fun getLabelById(id: Long): LabelResponseTo {
        val label = labelRepository.findById(id).orElseThrow { NoSuchElementException() }
        return label.toResponse()
    }

    fun getAllLabels(): List<LabelResponseTo> =
        labelRepository.findAll().map { it.toResponse() }

    fun updateLabel(id: Long, labelRequestTo: LabelRequestTo): LabelResponseTo {
        val updatedLabel = labelRequestTo.toEntity().apply { this.id = id }
        return labelRepository.save(updatedLabel).toResponse()
    }

    fun deleteLabel(id: Long) {
        labelRepository.findById(id).orElseThrow { NoSuchElementException() }
        labelRepository.deleteById(id)
    }

    fun getLabelsByIssueId(issueId: Long): List<LabelResponseTo> {
        val issue = issueRepository.findById(issueId).orElseThrow { NoSuchElementException() }
        return issue.labels.map { it.toResponse() }
    }
}